io/zenoh-transport/src/unicast/universal/tx.rs

```diff
@@ -11,51 +11,56 @@
 // Contributors:
 //   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
 //
-use zenoh_core::zread;
-use zenoh_protocol::network::NetworkMessage;
+use zenoh_protocol::{core::Reliability, network::NetworkMessage};
 
 use super::transport::TransportUnicastUniversal;
 #[cfg(feature = "shared-memory")]
 use crate::shm::map_zmsg_to_partner;
+use crate::unicast::transport_unicast_inner::TransportUnicastTrait;
 
 impl TransportUnicastUniversal {
     fn schedule_on_link(&self, msg: NetworkMessage) -> bool {
-        macro_rules! zpush {
-            ($guard:expr, $pipeline:expr, $msg:expr) => {
-                // Drop the guard before the push_zenoh_message since
-                // the link could be congested and this operation could
-                // block for fairly long time
-                let pl = $pipeline.clone();
-                drop($guard);
-                tracing::trace!("Scheduled: {:?}", $msg);
-                return pl.push_network_message($msg);
-            };
-        }
+        let transport_links = self
+            .links
+            .read()
+            .expect("reading `TransportUnicastUniversal::links` should not fail");
 
-        let guard = zread!(self.links);
-        // First try to find the best match between msg and link reliability
-        if let Some(pl) = guard.iter().find_map(|tl| {
-            if msg.is_reliable() == tl.link.link.is_reliable() {
-                Some(&tl.pipeline)
-            } else {
-                None
-            }
-        }) {
-            zpush!(guard, pl, msg);
-        }
+        let msg_reliability = Reliability::from(msg.is_reliable());
 
-        // No best match found, take the first available link
-        if let Some(pl) = guard.iter().map(|tl| &tl.pipeline).next() {
-            zpush!(guard, pl, msg);
-        }
+        let Some(transport_link) = transport_links
+            .iter()
+            .find(|transport_link| {
+                transport_link.link.config.reliability == msg_reliability
+                    && transport_link.link.config.priority == Some(msg.priority())
```

Comment:
This is a strict comparison, I believe a better strategy would be finding the link with closest priority matching. 
Moreover, there are too many calls to `find` which iterates over the list of the links every time. Wouldn't be better to just iterate once?

Reply:
What do you mean by "closest priority matching"? If the message data specified priority 3 and we have two linked tagged with priorities 2 and 4, then which link do we choose?

Rounding down will mean that the links with lower priorities will get more of the messages for which there is no exact match. And vice-versa for rounding up.

I'll remove the extra loops.

Reply:
I'll go with rounding-up. It's less problematic impacting data with lower priority than higher priority.

Reply:
As we're using ranges now, I interpreted "closest priority matching" to mean belonging to the smallest range available. Please let me know what you think about that.